In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

df = pd.read_csv('../../data/bronze/articles.csv')

In [4]:
print("Shape: ", df.shape)

Shape:  (105542, 25)


In [5]:
df.head()

,article_id,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,...,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,108775015,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,9,Black,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
1,108775044,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,10,White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
2,108775051,108775,Strap top (1),253,Vest top,Garment Upper body,1010017,Stripe,11,Off White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
3,110065001,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,9,Black,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."
4,110065002,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,10,White,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105542 entries, 0 to 105541
Data columns (total 25 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   article_id                    105542 non-null  int64 
 1   product_code                  105542 non-null  int64 
 2   prod_name                     105542 non-null  object
 3   product_type_no               105542 non-null  int64 
 4   product_type_name             105542 non-null  object
 5   product_group_name            105542 non-null  object
 6   graphical_appearance_no       105542 non-null  int64 
 7   graphical_appearance_name     105542 non-null  object
 8   colour_group_code             105542 non-null  int64 
 9   colour_group_name             105542 non-null  object
 10  perceived_colour_value_id     105542 non-null  int64 
 11  perceived_colour_value_name   105542 non-null  object
 12  perceived_colour_master_id    105542 non-null  int64 
 13 

In [7]:
df.describe()

,article_id,product_code,product_type_no,graphical_appearance_no,colour_group_code,perceived_colour_value_id,perceived_colour_master_id,department_no,index_group_no,section_no,garment_group_no
count,1.055420e+05,105542.000000,105542.000000,1.055420e+05,105542.000000,105542.000000,105542.000000,105542.000000,105542.000000,105542.000000,105542.000000
mean,6.984246e+08,698424.563378,234.861875,1.009515e+06,32.233822,3.206183,7.807972,4532.777833,3.171534,42.664219,1010.438290
std,1.284624e+08,128462.384432,75.049308,2.241359e+04,28.086154,1.563839,5.376727,2712.692011,4.353234,23.260105,6.731023
min,1.087750e+08,108775.000000,-1.000000,-1.000000e+00,-1.000000,-1.000000,-1.000000,1201.000000,1.000000,2.000000,1001.000000
25%,6.169925e+08,616992.500000,252.000000,1.010008e+06,9.000000,2.000000,4.000000,1676.000000,1.000000,20.000000,1005.000000
50%,7.022130e+08,702213.000000,259.000000,1.010016e+06,14.000000,4.000000,5.000000,4222.000000,2.000000,46.000000,1009.000000
75%,7.967030e+08,796703.000000,272.000000,1.010016e+06,52.000000,4.000000,11.000000,7389.000000,4.000000,61.000000,1017.000000
max,9.594610e+08,959461.000000,762.000000,1.010029e+06,93.000000,7.000000,20.000000,9989.000000,26.000000,97.000000,1025.000000


In [8]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
article_id,105542.0,6.984246e+08,1.284624e+08,108775015.0,616992501.0,702213001.5,7.967030e+08,959461001.0
product_code,105542.0,6.984246e+05,1.284624e+05,108775.0,616992.5,702213.0,7.967030e+05,959461.0
product_type_no,105542.0,2.348619e+02,7.504931e+01,-1.0,252.0,259.0,2.720000e+02,762.0
graphical_appearance_no,105542.0,1.009515e+06,2.241359e+04,-1.0,1010008.0,1010016.0,1.010016e+06,1010029.0
colour_group_code,105542.0,3.223382e+01,2.808615e+01,-1.0,9.0,14.0,5.200000e+01,93.0
perceived_colour_value_id,105542.0,3.206183e+00,1.563839e+00,-1.0,2.0,4.0,4.000000e+00,7.0
perceived_colour_master_id,105542.0,7.807972e+00,5.376727e+00,-1.0,4.0,5.0,1.100000e+01,20.0
department_no,105542.0,4.532778e+03,2.712692e+03,1201.0,1676.0,4222.0,7.389000e+03,9989.0
index_group_no,105542.0,3.171534e+00,4.353234e+00,1.0,1.0,2.0,4.000000e+00,26.0
section_no,105542.0,4.266422e+01,2.326010e+01,2.0,20.0,46.0,6.100000e+01,97.0


In [9]:
print(df.columns.tolist())

['article_id', 'product_code', 'prod_name', 'product_type_no', 'product_type_name', 'product_group_name', 'graphical_appearance_no', 'graphical_appearance_name', 'colour_group_code', 'colour_group_name', 'perceived_colour_value_id', 'perceived_colour_value_name', 'perceived_colour_master_id', 'perceived_colour_master_name', 'department_no', 'department_name', 'index_code', 'index_name', 'index_group_no', 'index_group_name', 'section_no', 'section_name', 'garment_group_no', 'garment_group_name', 'detail_desc']


In [10]:
(df == -1).sum()

article_id                        0
product_code                      0
prod_name                         0
product_type_no                 121
product_type_name                 0
product_group_name                0
graphical_appearance_no          52
graphical_appearance_name         0
colour_group_code                28
colour_group_name                 0
perceived_colour_value_id        28
perceived_colour_value_name       0
perceived_colour_master_id      685
perceived_colour_master_name      0
department_no                     0
department_name                   0
index_code                        0
index_name                        0
index_group_no                    0
index_group_name                  0
section_no                        0
section_name                      0
garment_group_no                  0
garment_group_name                0
detail_desc                       0
dtype: int64

In [11]:
df.nunique()

article_id                      105542
product_code                     47224
prod_name                        45875
product_type_no                    132
product_type_name                  131
product_group_name                  19
graphical_appearance_no             30
graphical_appearance_name           30
colour_group_code                   50
colour_group_name                   50
perceived_colour_value_id            8
perceived_colour_value_name          8
perceived_colour_master_id          20
perceived_colour_master_name        20
department_no                      299
department_name                    250
index_code                          10
index_name                          10
index_group_no                       5
index_group_name                     5
section_no                          57
section_name                        56
garment_group_no                    21
garment_group_name                  21
detail_desc                      43404
dtype: int64

In [12]:
cols_with_minus1 = ['product_type_no', 'graphical_appearance_no',
                    'colour_group_code', 'perceived_colour_value_id',
                    'perceived_colour_master_id']
for col in cols_with_minus1:
    print(f"Column {col}: {(df[col] == -1).sum()}")

Column product_type_no: 121
Column graphical_appearance_no: 52
Column colour_group_code: 28
Column perceived_colour_value_id: 28
Column perceived_colour_master_id: 685


In [13]:
print(df.groupby('product_type_no')['product_type_name'].nunique().max())
print(df.groupby('colour_group_code')['colour_group_name'].nunique().max())

1
1


In [17]:
print(df[df['product_type_no'] == -1]['product_type_name'].value_counts())

product_type_name
Unknown    121
Name: count, dtype: int64


In [20]:
print(df[df['graphical_appearance_no'] == -1]['graphical_appearance_name'].value_counts())

graphical_appearance_name
Unknown    52
Name: count, dtype: int64


In [21]:
print(df[df['colour_group_code'] == -1]['colour_group_name'].value_counts())

colour_group_name
Unknown    28
Name: count, dtype: int64


In [23]:
print(df[df['perceived_colour_value_id'] == -1]['perceived_colour_value_name'].value_counts())
print(df[df['perceived_colour_master_id'] == -1]['perceived_colour_master_name'].value_counts())

perceived_colour_value_name
Unknown    28
Name: count, dtype: int64
perceived_colour_master_name
Unknown    685
Name: count, dtype: int64


## EDA Notes — Missing Values

All columns below contain -1 values where corresponding _name = "Unknown" → **drop rows**

| Column | -1 Count |
|---|---|
| product_type_no | 121 |
| graphical_appearance_no | 52 |
| colour_group_code | 28 |
| perceived_colour_value_id | 28 |
| perceived_colour_master_id | 685 |

In [26]:
paired_cols = [
    ('product_type_no', 'product_type_name'),
    ('graphical_appearance_no', 'graphical_appearance_name'),
    ('colour_group_code', 'colour_group_name'),
    ('perceived_colour_value_id', 'perceived_colour_value_name'),
    ('perceived_colour_master_id', 'perceived_colour_master_name'),
    ('index_group_no', 'index_group_name'),
    ('section_no', 'section_name'),
    ('garment_group_no', 'garment_group_name'),
    ('department_no', 'department_name'),
]

for no_col, name_col in paired_cols:
    print(f"{no_col}: {df[no_col].nunique()} unique | {name_col}: {df[name_col].nunique()} unique")

product_type_no: 132 unique | product_type_name: 131 unique
graphical_appearance_no: 30 unique | graphical_appearance_name: 30 unique
colour_group_code: 50 unique | colour_group_name: 50 unique
perceived_colour_value_id: 8 unique | perceived_colour_value_name: 8 unique
perceived_colour_master_id: 20 unique | perceived_colour_master_name: 20 unique
index_group_no: 5 unique | index_group_name: 5 unique
section_no: 57 unique | section_name: 56 unique
garment_group_no: 21 unique | garment_group_name: 21 unique
department_no: 299 unique | department_name: 250 unique


In [27]:
df.groupby('department_name')['department_no'].nunique().sort_values(ascending=False).head(10)

department_name
Knitwear             11
Trouser               7
Shorts                5
Jersey                5
Shirt                 4
Accessories           4
Blouse                3
Jersey Fancy          3
Dress                 3
Small Accessories     2
Name: department_no, dtype: int64

In [28]:
df.groupby('product_type_name')['product_type_no'].nunique().sort_values(ascending=False).head(5)

product_type_name
Umbrella           2
Accessories set    1
Sarong             1
Soft Toys          1
Socks              1
Name: product_type_no, dtype: int64

In [29]:
df.groupby('section_name')['section_no'].nunique().sort_values(ascending=False).head(5)

section_name
Ladies Other                     2
Baby Essentials & Complements    1
Men Accessories                  1
Men Edition                      1
Men H&M Sport                    1
Name: section_no, dtype: int64

## EDA Notes — Redundant Columns

Keep `_no` / `_id` columns, drop `_name` equivalents.
`department_no` more granular (299 vs 250 unique) → drop `department_name`

`prod_name`, `detail_desc` → drop

In [30]:
df.dtypes

article_id                       int64
product_code                     int64
prod_name                       object
product_type_no                  int64
product_type_name               object
product_group_name              object
graphical_appearance_no          int64
graphical_appearance_name       object
colour_group_code                int64
colour_group_name               object
perceived_colour_value_id        int64
perceived_colour_value_name     object
perceived_colour_master_id       int64
perceived_colour_master_name    object
department_no                    int64
department_name                 object
index_code                      object
index_name                      object
index_group_no                   int64
index_group_name                object
section_no                       int64
section_name                    object
garment_group_no                 int64
garment_group_name              object
detail_desc                     object
dtype: object

## EDA Notes — Column Types

These columns are int64 but categorical (code/ID nature, no mathematical meaning):
product_type_no, graphical_appearance_no, colour_group_code,
perceived_colour_value_id, perceived_colour_master_id,
department_no, index_group_no, section_no, garment_group_no

→ Convert to category type in cleaning step

In [34]:
cat_cols = [
    'product_type_no', 'graphical_appearance_no', 'colour_group_code',
    'perceived_colour_value_id', 'perceived_colour_master_id',
    'department_no', 'index_group_no', 'section_no', 'garment_group_no'
]

for col in cat_cols:
    top = df[col].value_counts(normalize=True).iloc[0] * 100
    print(f"{col}: most dominant category %{top:.1f}")

product_type_no: most dominant category %10.6
graphical_appearance_no: most dominant category %47.1
colour_group_code: most dominant category %21.5
perceived_colour_value_id: most dominant category %40.5
perceived_colour_master_id: most dominant category %21.4
department_no: most dominant category %1.9
index_group_no: most dominant category %37.7
section_no: most dominant category %6.9
garment_group_no: most dominant category %20.3


## EDA Notes — Distribution Analysis

No low variance issues (no column dominates >90%)
graphical_appearance_no: top category = 47.1% → monitor in modeling
All categorical columns are suitable for modeling

In [38]:
df["index_group_no"].value_counts()

index_group_no
1     39737
4     34711
2     15149
3     12553
26     3392
Name: count, dtype: int64

In [39]:
print(df[df["index_group_no"] == 26]["index_group_name"].value_counts())

index_group_name
Sport    3392
Name: count, dtype: int64
